engine-nextgen/src/main/kotlin/graphql/nadel/enginekt/transform/query/NadelQueryTransformer.kt

```diff
@@ -36,145 +63,113 @@ class NadelQueryTransformer internal constructor(
     ): List<ExecutableNormalizedField> {
-        val continuation = object : Continuation {
-            override suspend fun transform(fields: List<ExecutableNormalizedField>): List<ExecutableNormalizedField> {
-                return transformFields(
-                    executionContext = executionContext,
-                    transformContext = transformContext,
-                    service = service,
-                    fields = fields,
-                    executionPlan = executionPlan,
-                )
-            }
+        return fields.flatMap {
+            transform(it)
         }
+    }
 
+    suspend fun transform(
+        field: ExecutableNormalizedField,
+    ): List<ExecutableNormalizedField> {
         val transformationSteps: List<NadelExecutionPlan.Step<Any>> = executionPlan.transformationSteps[field]
             ?: return listOf(
-                patchObjectTypeNames(service, field)
-                    .children(continuation.transform(field.children))
-                    .build()
-                    .also { newField ->
-                        // Track overall -> underlying fields
-                        transformContext.overallToUnderlyingFields.compute(field) { _, oldValue ->
-                            (oldValue ?: emptyList()) + newField
-                        }
-                    }
+                transformPlain(field)
             )
 
-        val transformResult =
-            applyTransformationSteps(field, transformationSteps, executionContext, continuation, service)
+        return transform(field, transformationSteps)
+    }
+
+    private suspend fun transform(
+        field: ExecutableNormalizedField,
+        transformationSteps: List<NadelExecutionPlan.Step<Any>>,
+    ): List<ExecutableNormalizedField> {
+        val transformResult = applyTransformationSteps(field, transformationSteps)
 
-        // TODO: I think that patching names here is wrong, we're giving mixed signals to the transformations
-        // i.e. yes please give us a field in the underlying sense, but only field name
-        // I think this introduces confusion
-        val patchedArtificialFields = patchObjectTypeNames(service, transformResult.artificialFields)
-        val patchedNewField = patchObjectTypeNames(service, listOfNotNull(transformResult.newField))
-        transformContext.artificialFields.addAll(patchedArtificialFields)
+        val artificialFields = transformResult.artificialFields.map {
+            it.toBuilder()
+                .clearObjectTypesNames()
+                .objectTypeNames(getUnderlyingTypeNames(it.objectTypeNames))
+                .build()
+        }
+
+        val newField = listOfNotNull(
+            transformResult.newField?.let {
+                it.toBuilder()
+                    .clearObjectTypesNames()
+                    .objectTypeNames(getUnderlyingTypeNames(it.objectTypeNames))
+                    .children(transform(it.children))
```

Comment:
Ok so this is the actual fix. Previously we would make the transforms themselves do this.

Issue is, say you have this

```
class AriArgumentTransform {
  fun transformField() {
    return Result(
      field = field.toBuilder()
        .arguments(…) // Interpret ARIs
        .children(transformer.transform(field.children)) // You NEEDED this
        .build()
    )
  }
}
```

We MUST call `transformer.transform` on children because the children can have renames, hydrations etc.

BUT what happens if the `AriArgumentTransform` is NOT the only transform on the field. What if the field with an ARI argument is renamed. Then we have BOTH the `AriArgumentTransform` AND `RenameTransform` transforming the children. Which is not good.